File: 2020/day15.hs
import Data.Map.Strict (Map, (!?), fromList, insert)

main :: IO ()
main = do
  let input = [2, 1, 10, 11, 0, 6]
      mem = fromList $ zip (init input) [1..]
      numbers = 0 : init input ++ genNumList mem (length input) (last input)
      genNumList m i n = seq n
        $ n : genNumList (insert n i m) (i + 1) (maybe 0 (i -) $ m !? n)

  putStrLn . ("Part 1: " ++) . show $ numbers !! 2020
  putStrLn . ("Part 2: " ++) . show $ numbers !! 30000000

File: 2020/day15.lua
input = {[2] = 1, [1] = 2, [10] = 3, [11] = 4, [0] = 5}
orig_len = 6
current_num = 6

for i = 1, 30000000 - orig_len do
  prev_index = input[current_num]
  input[current_num] = orig_len + i - 1
  current_num = prev_index and (orig_len + i - 1 - prev_index) or 0
  if i == 2020 - orig_len then print("Part 1:", current_num) end
end

print("Part 2:", current_num)
